Add `newlines-between` option to `order` rule

singles opened the pull request:
Checks existence of new lines between and inside import groups.

Implements #282.

benmosher commented:
LGTM, but @jfmengels should have final say, it's his baby. 😊

benmosher commented:
Oh, actually, it would be great if you could add a note to the change log. `### Added` section makes sense to me.

jfmengels commented on src/rules/order.js:
```diff
@@ -109,6 +109,37 @@ function convertGroupsToRanks(groups) {
+          previousImport.node, 'There should be one empty line between import groups'
+        )
+      } else if (currentImport.rank === previousImport.rank
+        && getLineDifference(currentImport, previousImport) === 2)
```
`=== 2` --> `>= 2`

jfmengels commented on docs/rules/order.md:
```diff
@@ -89,3 +91,56 @@ You can set the options like this:
+
+### `newlines-between: [bool]`:
+
+Checks existence of new lines between and inside import groups. For sake of examples let's assume `groups` settings like:
```
I'm just proposing a new rewording here:
```
Enforces or forbids new lines between import groups.
If omitted, assertion messages will be neither enforced nor forbidden.
If set to true, a new line between each group will be enforced, and new lines inside a group will be forbidden.
If set to false, no new lines are allowed in the entire import section.
```

Btw, starting to think that having a string is more explicit (readable in the config), and more ESLint-like. `newlines-between: "always|never".

jfmengels commented on docs/rules/order.md:
```diff
@@ -89,3 +91,56 @@ You can set the options like this:
+
+**a)** Not setting `newlines-between` option will cause rule not to check any empty lines validity.
+
+**b)** With `newlines-between: true` this will be ok:
```
Do not set a new group setting for the example, let's keep it simple and keep the default setting.
Instead of creating a section for every possibility, set and show it in the example, like

```js
/* eslint import/order: ["error", {"newlines-between": true}] */
var fs = require('fs');
```

Maybe keep examples short? With like 2 groups and 3 imports (suggestion, do what you think renders best here)
And if you need to have comments, add a space after `// ` for consistency :)

jfmengels commented:
This looks pretty good @singles, thanks a lot for this!
Just some comments on the documentation mostly.

> it's his baby

Man, you have no idea... I still haven't had the heart to deprecate `eslint-plugin-import-order`. It's my most popular repo, and growing in downloads every day... Will do it soon... soon...

benmosher commented:
I can imagine. But every download here is your download too, now! 😁 I think GitHub itself may be responsible for the recent uptick in downloads, so that's pretty cool.

jfmengels commented:
I think https://github.com/dustinspecker/awesome-eslint is partially responsible too. Even more so probably.

singles commented:
OK, new changes:

1. Added info the changelog
2. Fixed https://github.com/benmosher/eslint-plugin-import/pull/298#discussion_r61781477
3. Changed option values from `true|false` to `always|never`
4. Updated docs, less words, more concise

Thx @benmosher @jfmengels for feedback!

jfmengels commented on CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@ This change log adheres to standards from [Keep a CHANGELOG](http://keepachangel
 ### Added
 - [`newline-after-import`], new rule. ([#245], thanks [@singles])
 - Added an `optionalDependencies` option to [`no-extraneous-dependencies`] to allow/forbid optional dependencies ([#266], thanks [@jfmengels]).
+- Added `newlines-between` option to `order` rule ([#298], thanks [@singles])
```
Could you put brackts around `order` so that there is a link there?

singles replied:
Fixed.

jfmengels commented on docs/rules/order.md:
```diff
@@ -89,3 +91,48 @@ You can set the options like this:
+- If set to `always`, a new line between each group will be enforced, and new lines inside a group will be forbidden.
+- If set to `never`, no new lines are allowed in the entire import section.
+
+For default group setting those will be valid:
```
With the default group setting, the following will be valid:

singles replied:
Fixed.

jfmengels commented on docs/rules/order.md:
````diff
@@ -89,3 +91,48 @@ You can set the options like this:
+var sibling = require('./foo');
+```
+
+while those will be invalid:
````
`while the following will be invalid:`

Actually, if you could swap both sections, the invalid example is always (afaik) put before the valid section.

singles replied:
No problem, done.

jfmengels commented on docs/rules/order.md:
```diff
@@ -89,3 +91,48 @@ You can set the options like this:
+/* eslint import/order: ["error", {"newlines-between": "always"}] */
+var fs = require('fs');
+
+var index = require('./');
```
Could you add
```js
var path = require('path');
```
otherwise it looks like you simply enforce a line between each import.

And since most examples in the repo us the import syntax, might be nice to use that instead?

singles replied:
Sure, good idea, done.

jfmengels commented:
Sorry for the nitpicking, looks great otherwise @singles!

singles commented:
> Sorry for the nitpicking

No problem :)

jfmengels commented:
Looks pretty great to me :D

jfmengels commented:
![](https://media.giphy.com/media/fM3xoVGOpR90Q/giphy.gif)
